modelling/src/main/java/org/axonframework/modelling/StateManager.java

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.modelling;
+
+import jakarta.annotation.Nonnull;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+import org.axonframework.modelling.repository.AsyncRepository;
+import org.axonframework.modelling.repository.ManagedEntity;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The {@code StateManager} enables applications to load entities based on the type of the entity and an id, and to
+ * persist them. Only one {@link AsyncRepository} or load and save function can be registered for a type, as this is
+ * considered their identity.
+ * <p>
+ * Entities are registered by their type, and can be registered by either:
+ * <ul>
+ *     <li>Registering an {@link AsyncRepository} for a certain type, or</li>
+ *     <li>Registering a load and save function for a certain type.</li>
+ * </ul>
+ * <p>
+ * In the latter case, the state manager will create a {@link SimpleEntityAsyncRepository} for the given type with
+ * the given load and save functions.
+ *
+ * @author Mitchell Herrijgers
+ * @since 5.0.0
+ */
+public interface StateManager {
+
+    /**
+     * Registers an {@link AsyncRepository} for state type {@code T} with id of type {@code I}. Only one
+     * {@link AsyncRepository} can be registered for a type.
+     *
+     * @param idType     The type of the identifier.
+     * @param stateType  The type of the state.
+     * @param repository The {@link AsyncRepository} to use for loading state.
+     * @param <I>        The type of id.
+     * @param <T>        The type of state.
+     * @return This instance for fluent interfacing.
+     */
+    <I, T> StateManager register(
+            @Nonnull Class<I> idType,
+            @Nonnull Class<T> stateType,
+            @Nonnull AsyncRepository<I, T> repository
+    );
+
+    /**
+     * Registers a load and save function for state type {@code T} with id of type {@code I}. Creates a
+     * {@link SimpleEntityAsyncRepository} for the given type with the given load and save functions.
+     *
+     * @param idType    The type of the identifier.
+     * @param stateType The type of the state.
+     * @param loader    The function to load state.
+     * @param persister The function to persist state.
+     * @param <I>       The type of id.
+     * @param <T>       The type of state.
+     * @return This instance for fluent interfacing.
+     */
+    default <I, T> StateManager register(
+            @Nonnull Class<I> idType,
+            @Nonnull Class<T> stateType,
+            @Nonnull SimpleEntityLoader<I, T> loader,
+            @Nonnull SimpleEntityPersister<I, T> persister
+    ) {
+        return register(idType, stateType, new SimpleEntityAsyncRepository<>(idType, stateType, loader, persister));
```

Comment:
Ah, I guess the purpose of the `SimpleEntityAsyncRepository` is to ensure the `StateManager` implementation only has to deal with repositories, while the configuration has freedom to provide separate `loader` and `persister` lambdas. Is that a fair assumption?

Reply:
Yes! :) It can also later be used to wrap other frameworks.